File: graphics/projector.py
"""Project one authoritative world snapshot into a browser RenderFrame."""
from __future__ import annotations
import copy
from typing import Any, Iterable
from world.catalog import MapCatalog
from .assets import entity_asset
from .camera import flat_camera, quantize_x, screen_x
from .contracts import SCHEMA_VERSION, validate_render_frame
from .presentation import terrain_layer

REST_VELOCITY = 0.05

class SceneProjector:
    def __init__(self, catalog: MapCatalog | None = None):
        self.catalog = catalog or MapCatalog.load_default()

    @staticmethod
    def _interaction_state(entity_id: str, interactions: Iterable[dict[str, Any]]) -> str | None:
        for item in interactions:
            if not isinstance(item, dict) or item.get("entity_id") != entity_id:
                continue
            if item.get("object_id") == "workstation" and item.get("interaction_id") in {"sit","work"} and item.get("status") in {"active","arrived","confirmed"}:
                return "seated_working"
        return None

    def project(self, snapshot: dict[str, Any], *, focus_entity_id: str,
                interactions: Iterable[dict[str, Any]] = (),
                authoritative: bool = True, source: str = "embodied_world") -> dict[str, Any]:
        required = {"world_id","world_epoch","world_tick","world_revision","entities"}
        if not isinstance(snapshot, dict) or not required <= set(snapshot):
            raise ValueError("world snapshot is incomplete")
        if type(snapshot["world_tick"]) is not int or snapshot["world_tick"] < 0:
            raise ValueError("world_tick is invalid")
        if type(snapshot["world_revision"]) is not int or snapshot["world_revision"] < 0:
            raise ValueError("world_revision is invalid")
        entities = snapshot["entities"]
        if not isinstance(entities, list):
            raise ValueError("entities must be a list")
        focus = next((item for item in entities if item.get("entity_id") == focus_entity_id), None)
        if focus is None:
            raise ValueError("focus entity is absent from world snapshot")
        zone_id = focus.get("zone_id")
        self.catalog.get(zone_id)
        camera = flat_camera(self.catalog.physics(zone_id))
        layer = terrain_layer(self.catalog, zone_id)
        rendered, seen = [], set()
        for entity in entities:
            if entity.get("zone_id") != zone_id:
                continue
            entity_id = entity.get("entity_id")
            if not isinstance(entity_id, str) or not entity_id or entity_id in seen:
                raise ValueError("snapshot entity IDs must be unique non-empty strings")
            seen.add(entity_id)
            x, vx = float(entity["x"]), float(entity.get("vx", 0.0))
            pose = self._interaction_state(entity_id, interactions)
            rendered.append({
                "entity_id":entity_id, "transform":{"x":x},
                "display_cell":quantize_x(x,camera), "screen_x":screen_x(x,camera),
                "visual_asset":entity_asset(entity),
                "animation_state":pose or ("moving" if abs(vx)>REST_VELOCITY else "idle"),
            })
        frame = {
            "schema_version":SCHEMA_VERSION,
            "frame_id":f"frame:{snapshot['world_epoch']}:{snapshot['world_revision']}:{zone_id}",
            "source_world_epoch":str(snapshot["world_epoch"]),
            "source_world_tick":snapshot["world_tick"],
            "source_world_revision":snapshot["world_revision"],
            "zone_id":zone_id, "camera":camera,
            "terrain_revision":layer["terrain_revision"],
            "entities":rendered, "props":copy.deepcopy(layer["objects"]),
            "presentation_revision":f"{zone_id}@{layer['map_version']}:{layer['theme_id']}",
            "freshness":{
                "state":"current","age_ticks":0,"authoritative":bool(authoritative),
                "source":source,"previous_epoch":snapshot.get("previous_epoch"),
            },
        }
        return validate_render_frame(frame)

    def terrain(self, zone_id: str) -> dict:
        return terrain_layer(self.catalog, zone_id)

    @staticmethod
    def layers(frame: dict[str, Any], terrain: dict[str, Any]) -> dict[str, Any]:
        validate_render_frame(frame)
        cells = list(terrain["cells"])
        if len(cells) != frame["camera"]["cells"]:
            raise ValueError("terrain and frame camera sizes differ")
        occupancy = [[] for _ in cells]
        for entity in frame["entities"]:
            occupancy[entity["display_cell"]].append(entity["entity_id"])
        return {"terrain":cells,"occupancy":occupancy}

__all__ = ["REST_VELOCITY","SceneProjector"]

File: player-gateway/src/projector.mjs
import fs from "node:fs";
import path from "node:path";
import {entityAsset, objectAsset} from "./assets.mjs";

const REST_VELOCITY = 0.05;
const MAX_FRAME_BYTES = 32 * 1024;

export function loadMapCatalog(mapDirectory) {
  const maps = new Map();
  for (const name of fs.readdirSync(mapDirectory).filter((item) => item.endsWith(".json")).sort()) {
    const value = JSON.parse(fs.readFileSync(path.join(mapDirectory, name), "utf8"));
    if (!value || typeof value.map_id !== "string" || !value.map_id) {
      throw new Error(`invalid map manifest: ${name}`);
    }
    if (maps.has(value.map_id)) throw new Error(`duplicate map_id: ${value.map_id}`);
    maps.set(value.map_id, value);
  }
  if (!maps.size) throw new Error("map catalog is empty");
  return maps;
}

function manifestFor(catalog, zoneId) {
  const value = catalog.get(zoneId);
  if (!value) throw new Error(`unknown map: ${zoneId}`);
  return value;
}

function cameraFor(manifest) {
  const bounds = manifest.physics?.bounds;
  const lo = Number(bounds?.x_min);
  const hi = Number(bounds?.x_max);
  if (!Number.isFinite(lo) || !Number.isFinite(hi) || lo >= hi) {
    throw new Error("map camera bounds are invalid");
  }
  return {
    axis: "x",
    world_min: lo,
    world_max: hi,
    cells: Math.trunc(hi - lo) + 1,
    projection: "linear",
  };
}

function quantizeX(x, camera) {
  const value = Number(x);
  if (!Number.isFinite(value) || value < camera.world_min || value > camera.world_max) {
    throw new Error("x is outside camera bounds");
  }
  return Math.min(camera.cells - 1, Math.floor(value - camera.world_min));
}

function screenX(x, camera) {
  const value = Number(x);
  if (!Number.isFinite(value) || value < camera.world_min || value > camera.world_max) {
    throw new Error("x is outside camera bounds");
  }
  return (value - camera.world_min) / (camera.world_max - camera.world_min);
}

export function terrainFor(catalog, zoneId) {
  const manifest = manifestFor(catalog, zoneId);
  const camera = cameraFor(manifest);
  const objects = (manifest.semantics?.objects || []).map((item) => ({
    object_id: item.object_id,
    kind: item.kind,
    x: Number(item.x),
    display_cell: quantizeX(item.x, camera),
    visual_asset: objectAsset(item.kind),
  }));
  return {
    map_id: zoneId,
    map_version: manifest.map_version,
    terrain_revision: manifest.presentation.terrain_revision,
    theme_id: manifest.presentation.theme_id,
    background_asset: "background." + manifest.presentation.theme_id,
    camera,
    base_asset: "terrain.road",
    objects,
  };
}

export function projectWorldSnapshot(snapshot, focusEntityId, catalog, {
  source = "player_gateway_v1",
} = {}) {
  if (!snapshot || typeof snapshot !== "object") throw new Error("world snapshot is required");
  for (const key of ["world_id", "world_epoch", "world_tick", "world_revision", "entities"]) {
    if (!(key in snapshot)) throw new Error("world snapshot is incomplete");
  }
  if (!Number.isInteger(snapshot.world_tick) || snapshot.world_tick < 0) {
    throw new Error("world_tick is invalid");
  }
  if (!Number.isInteger(snapshot.world_revision) || snapshot.world_revision < 0) {
    throw new Error("world_revision is invalid");
  }
  if (!Array.isArray(snapshot.entities)) throw new Error("entities must be an array");
  const focus = snapshot.entities.find((entity) => entity?.entity_id === focusEntityId);
  if (!focus) throw new Error("focus entity is absent from world snapshot");
  const zoneId = focus.zone_id;
  const manifest = manifestFor(catalog, zoneId);
  const camera = cameraFor(manifest);
  const terrain = terrainFor(catalog, zoneId);
  const seen = new Set();
  const entities = [];
  for (const entity of snapshot.entities) {
    if (entity?.zone_id !== zoneId) continue;
    const entityId = entity.entity_id;
    if (typeof entityId !== "string" || !entityId || seen.has(entityId)) {
      throw new Error("snapshot entity IDs must be unique non-empty strings");
    }
    seen.add(entityId);
    const x = Number(entity.x);
    const vx = Number(entity.vx || 0);
    if (!Number.isFinite(vx)) throw new Error("entity vx must be finite");
    entities.push({
      entity_id: entityId,
      transform: {x},
      display_cell: quantizeX(x, camera),
      screen_x: screenX(x, camera),
      visual_asset: entityAsset(entity),
      animation_state: Math.abs(vx) > REST_VELOCITY ? "moving" : "idle",
    });
  }
  const frame = {
    schema_version: 1,
    frame_id: `frame:${snapshot.world_epoch}:${snapshot.world_revision}:${zoneId}`,
    source_world_epoch: String(snapshot.world_epoch),
    source_world_tick: snapshot.world_tick,
    source_world_revision: snapshot.world_revision,
    zone_id: zoneId,
    camera,
    terrain_revision: terrain.terrain_revision,
    entities,
    props: terrain.objects,
    presentation_revision: `${zoneId}@${manifest.map_version}:${manifest.presentation.theme_id}`,
    freshness: {
      state: "current",
      age_ticks: 0,
      authoritative: true,
      source,
      previous_epoch: snapshot.previous_epoch ?? null,
    },
  };
  const size = Buffer.byteLength(JSON.stringify(frame));
  if (size > MAX_FRAME_BYTES) throw new Error(`RenderFrame exceeds ${MAX_FRAME_BYTES} bytes`);
  return frame;
}

export function projectHostState(state, catalog, options = {}) {
  const session = state?.session;
  const snapshot = state?.snapshot;
  const entityId = session?.entity_id;
  if (typeof entityId !== "string" || !entityId) throw new Error("Host session has no entity_id");
  return projectWorldSnapshot(snapshot, entityId, catalog, options);
}

export {MAX_FRAME_BYTES, REST_VELOCITY};
